src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs

```diff
@@ -114,36 +117,89 @@ protected override Metadata RequestHeadersCore
-                foreach (var entry in rpcException.Trailers)
+                ProcessHandlerError(ex, method);
+                return Task.CompletedTask;
+            }
+
+            var lockTask = _deadlineLock.WaitAsync();
+            if (lockTask.IsCompletedSuccessfully)
+            {
+                try
                 {
-                    ResponseTrailers.Add(entry);
+                    ProcessHandlerError(ex, method);
+                }
+                finally
+                {
+                    _deadlineLock.Release();
                 }
 
-                _status = rpcException.Status;
+                return Task.CompletedTask;
             }
             else
             {
-                Log.ErrorExecutingServiceMethod(_logger, method, ex);
+                return ProcessHandlerErrorAsyncCore(lockTask, ex, method);
+            }
+
+            async Task ProcessHandlerErrorAsyncCore(Task lockTask, Exception ex, string method)
+            {
+                await lockTask;
 
-                var message = ErrorMessageHelper.BuildErrorMessage("Exception was thrown by handler.", ex, ServiceOptions.EnableDetailedErrors);
-                _status = new Status(StatusCode.Unknown, message);
+                try
+                {
+                    ProcessHandlerError(ex, method);
+                }
+                finally
+                {
+                    _deadlineLock!.Release();
+                }
             }
+        }
 
-            HttpContext.Response.ConsolidateTrailers(this);
+        private void ProcessHandlerError(Exception ex, string method)
+        {
+            try
+            {
+                if (ex is RpcException rpcException)
+                {
+                    Log.RpcConnectionError(_logger, rpcException.StatusCode, ex);
+
+                    // There are two sources of metadata entries on the server-side:
+                    // 1. serverCallContext.ResponseTrailers
+                    // 2. trailers in RpcException thrown by user code in server side handler.
+                    // As metadata allows duplicate keys, the logical thing to do is
+                    // to just merge trailers from RpcException into serverCallContext.ResponseTrailers.
+                    foreach (var entry in rpcException.Trailers)
+                    {
+                        ResponseTrailers.Add(entry);
+                    }
+
+                    _status = rpcException.Status;
+                }
+                else
+                {
+                    Log.ErrorExecutingServiceMethod(_logger, method, ex);
+
+                    var message = ErrorMessageHelper.BuildErrorMessage("Exception was thrown by handler.", ex, ServiceOptions.EnableDetailedErrors);
+                    _status = new Status(StatusCode.Unknown, message);
+                }
+
+                // Don't update trailers if request has exceeded deadline/aborted
+                if (!CancellationToken.IsCancellationRequested)
```

Comment:
This is not sufficient thread safety, you'd need a lock here.

Reply:
What race are you worried about? It looks like we should have the _deadlineLock here in ProcessHandlerError and in DeadlineExceeded where HttpContext.Abort() is called.

Reply:
Ah, nevermind if this is always called within the _deadlineLock.